### ISSUE TITLE
Support materials

### ISSUE THREAD

**Issue by vinivendra** [KillzXGaming/NTT-Model-Dumper#2]
Hey, I'm really interested in having materials support for this tool.

If you have any plans of adding that, please let me know. If not, I'd be happy to contribute, though there's a lot I don't understand here, so I'd probably need some help.

From my basic understanding of the DAE format, we'd need both a list of all materials used in the file and a material associated with each vertex, right? Reading [this comment](https://github.com/KillzXGaming/NTT-Model-Dumper/blob/c4f8749d5a20c0347b299f822d486322886818b3/Model.cs#L106) it seems like the basic structure is already there for the first part, so that's probably where I'd start if I were to do this myself.

Thank you for making this tool in the first place, and for any help you can give me :)

**Comment by vinivendra** [comment]
OK, I read the code and the Collada specification and I think I understand most of what's going on here. I managed to add the list of materials in the file, but I haven't yet found how the materials are associated to each mesh or vertex.

Looking into the .model files, the only mentions of the material's paths or names I found were in the beginning of the file. This makes me think that the meshes reference the materials by index (instead of by name) which makes it harder to know where these indices are.

@KillzXGaming, do you have any idea how (or where) these materials are referenced?

**Comment by KillzXGaming** [comment]
I haven't figured out how they are referenced but I think it might be at the very bottom of the file. There is a sort of node tree with all the meshes.

As for the material data itself, some of the material data can be seen as raw text. It is like a sort of shader node tree.

![image](https://user-images.githubusercontent.com/13475262/165624933-0fccfdb3-dbd0-4843-939e-2760ce87ed28.png)

Might be difficult to support if materials vary alot between models.

**Comment by vinivendra** [comment]
Sorry, I don't understand if you talking about the .model or the .dae file... what's the file from your image? I'm currently looking at the `DREADNOUGHTHANGAR_PILLARA_MERGED_DX11.MODEL` and the exported DAE, but I can't find anything like that.

The end of my .model file (after the last DXTV section) has like the ASCII alphabet four times, then four sections called `Combined_ROOT`, `Combined_LOD1`, `Combined_SHAD` and `Combined_ZOCC`, but I'm not sure if this is what you're talking about.

Just to be clear, I'm not yet worried about how to export the material's shading information into DAE. I wanted first to export the amount of materials and their names (which seems to be OK) and what vertices use each material (which is what I can't find in the .model file).

**Comment by KillzXGaming** [comment]
Yeah the section with those names, Combined_ROOT, Combined_LOD1, etc in .model . I haven't figured out the structure of those yet but it contains alot of data.

That shader node data is just info in the .material file.

**Comment by vinivendra** [comment]
OK, so I investigated a bit more and it looks like the MODEL files already have the meshes separated by material (probably why there are so many meshes), and that the meshes are in the same order as the materials show up in the files.

The only thing that messes up a simple 1-to-1 correspondence between meshes and materials are the duplicated meshes for higher/lower levels of detail.

For example, in the `DREADNOUGHTHANGAR_PILLARA_MERGED_DX11.MODEL` file we have the following meshes:

<img width="920" alt="image" src="https://user-images.githubusercontent.com/1630012/166158187-d0cb9647-5839-4097-9ac3-184e239d6027.png">

I checked with the materials and textures (and how it looks in the game) and this makes sense:

- Mesh 1 uses material 1.
- Mesh 2 uses material 2.
- Mesh 3 uses material 3.
- Mesh 4 is a LOD variation on mesh 1, and therefore uses material 1.
- Mesh 5 is a LOD variation on mesh 2, and therefore uses material 2.
- Mesh 6 is a LOD variation on mesh 3, and therefore uses material 3.
- Mesh 7 uses the next free material, material 4.
- And mesh 8 uses material 5.

These are meshes 1, 2 and 3 together, using the textures from each material as normal maps, as compared to the game:

<img width="451" alt="Screen Shot 2022-05-01 at 15 06 15" src="https://user-images.githubusercontent.com/1630012/166158702-e41bd2f4-cbff-4485-9869-04eb38bec360.png">

<img width="324" alt="Screen Shot 2022-05-01 at 15 00 58" src="https://user-images.githubusercontent.com/1630012/166158573-8ae0bc8f-f3ea-4df3-bfbc-30dd1c05753c.png">

I also looked at a few other meshes, including `DEATHSTARTWO_DX11.MODEL`, which was much more complicated but seemed to follow the same idea. There are 18 meshes and 17 materials in it, and the materials correspond to each mesh in order, except for one LOD variation on one of the last meshes.

If all this is right, we should be able to assign materials to meshes once we figure out how the LOD variations are specified. The thing is, I have no idea how to do that.

The pillar model has one `Combined_LOD1` section and 3 LOD meshes; the death star 2 model has a `Combined_LOD1` and a `Combined_LOD2` sections for (as far as I can tell) 1 LOD mesh.

I tried looking at each section and they seem to have a common structure. Here is (from left to right) Death Star 2's `Combined_LOD1`, `Combined_LOD2`, and pillar's `Combined_LOD1`:

<img width="1381" alt="Screen Shot 2022-05-01 at 15 13 57" src="https://user-images.githubusercontent.com/1630012/166158967-604536f4-7fd0-4aa6-a04e-fde8bd69a4ed.png">
<img width="1257" alt="Screen Shot 2022-05-01 at 15 14 15" src="https://user-images.githubusercontent.com/1630012/166158970-349f807b-f72f-4b93-b888-3d635e1e260b.png">

(The `<`s are just my notes)

That's where I'm stumped now though. I have no experience with reverse-engineering these hex files, and I look at them and can't imagine how they would reference the LOD meshes. I thought I'd post my thoughts here to see if anyone can spot something I missed, or even prove me wrong 😅

**Comment by vinivendra** [comment]
So I didn't make any progress on this specific issue, but I have some more info that might help future efforts at exporting materials.

First, I made a small script that turns the node graph in the `.MATERIAL` files into an SVG, which allows for better visualization of what's happening in it. I imagine the nodes mean something specific to the TSS shader, but I bet many of them could be replicated relatively easily in (for example) Blender's node graphs once we figure out a few kinks.
Here's the node graph for `ASSETS/MODELS/TILESETS/IMPERIAL/IMPERIAL_FLOORS_01.MATERIAL`:

![svg](https://user-images.githubusercontent.com/1630012/168450939-4316c551-c4bd-490d-b8e8-da0adce3d42f.svg)

Second, when trying to replicate some of these nodes I saw that the normals looked weird in the model. A bit of research later made me understand that it's because the files represent derivative maps, not the tangent-space normal maps I was expecting. That's why they're yellowish instead of blueish.

Here's `ASSETS/MODELS/TILESETS/IMPERIAL/TEXTURES/IMPERIAL_FLOORS_N_DX11.TEXTURE`:

![IMPERIAL_FLOORS_N_DX11](https://user-images.githubusercontent.com/1630012/168450998-960b7ad5-b261-459c-82f1-99ba5850d3dc.png)
